avalon/tools/projectmanager/widget.py

```diff
@@ -143,7 +143,18 @@ def preserve_selection(tree_view,
 
 def _list_project_silos():
     """List the silos from the project's configuration"""
-    silos = io.distinct("silo")
+    # Backwards compatible way to get silos/top assets from DB:
```

Comment:
Wouldn't it be easier to do this:

```python
silos = io.distinct("silo")   # Backwards compatible silo support

# New style silos getting assets without parent and silo
new_silos = [s['name'] for s in io.find({
        'type': 'asset',
        'data.visualParent': None,
        '$or': [{'silo': {'$exists': False}}, {'silo': None}]
    })]
silos.extend(new_silos)

return silos
```

- It supports backwards compatible silos together with new style silos in one project.
- It does less database queries.
- It's more readable (to me at least)

Reply:
That won't work actually because io.distinct("silo") will contain None in new way. I had it similar way before:
```
silos = [s['name'] for s in io.find({"type": "asset", "silo": None})]
silos_old = io.distinct("silo")
for silo in silos_old:
     if silo not in silos and silo is not None:
         silos.append(silo)
```
But the point of this commit was prepare to not use 'silo' column in queries at all

Reply:
Do we have to change `silo` at all? I thought this change was about ignoring silo, and using the top-level asset instead?

Reply:
> That won't work actually because io.distinct("silo") will contain None in new way.

Then like this?

```python
silos = io.distinct("silo") or []
```

Reply:
> Do we have to change `silo` at all? I thought this change was about ignoring silo, and using the top-level asset instead?

that's exactly what we're trying to avoid. I'll try to explain what @iLLiCiTiT was trying to achieve here :).

The key point here is how to make it backwards compatible while also making sure that the new way of determining what is the 'top-level asset' doesn't need any silo data whatsoever. 

I think the confusion here with the naming. Silo is just data, but what we're trying to do is make a list of Tabs in the GUI. This can be made either from silo data if the project has it, or finding assets with `visualParent == None`. The problem is that old project also have asset with visualParent None, but in those case, this project shouldn't become a Tab. So we actually need to distinguish between old type and new type of project to ensure backward compatibility. 

here's slight change to code that should make the distinction clear. This way one wouldn't need silo at all and its clear we're not actually collecting silos, just data that forms tab in the GUI.

```
def _list_project_silos():
    # Let's find whether we're dealing with old or new type of project (having no silos):
    silo_existence = {
        'type': 'asset',
        'silo': {'$exists': False}]
    }
    # if we have new project (one with no silo data at all), get tabs from visualParent == None
    if len(list(io.find(silo_existence))) > 0:
        tabs = [s['name'] for s in io.find({
            'type': 'asset',
            'data.visualParent': None
        })]
    # if we have old project get tabs the old way
    else:
        tabs = io.distinct("silo")

    if not silos:
        project = io.find_one({"type": "project"})
    return list(sorted(tabs))
```

Reply:
You wouldn't need to check for the full project having one of it. Whenever something does *not* have a visual parent and does *not* have silo data, then it's your top most silo tab - it's as simple as that, right? ;)

It will even allow compatibility inside a single project to mix it up. (if you work on older projects)

Reply:
And what about:
```
silos = io.distinct("silo")

# New style, works if 'silo' column doesn't exists
if not silos:
        silos = [s['name'] for s in io.find({
                'type': 'asset',
                'data.visualParent': None
            })]
```
- easy to separate, strictly dividing
- it doesn't allow to mix it up but I don't think it's good idea to allow mix different hierarchy approache, could cause huge issues

Reply:
> it doesn't allow to mix it up but I don't think it's good idea to allow mix different hierarchy approache, could cause huge issues

I'm actually not seeing something that could cause a big issue. ;) Even if you have a silo `film` and later you create that asset `film` that can act as silo (having no silo set and no visualParent) then I believe it would know how to parent it. Right?

I'm just more afraid of someone accidentally making an old style silo (or the other way around, a new one) and it not being picked up or suddenly discarding the ones you had. But yes, this is definitely already looking like more readable code.

Reply:
>I'm actually not seeing something that could cause a big issue

the main on is a bit down the line. Distinguishing what to display as the top asset in the hierachy widget appearing under the tabs. If you have old style silos, the top asset in the tree view is the one with visualParent:None, if you're however using that as tab , it should be the next one in the hierarchy. So if you allow mixed type in the project, it goes a bit down the rabit hole. Isn't it more trouble than it's worth? 

If we definitely want to allow mixed projects, then you're latest suggestion is the way to go. And we'd have to implement it on a few more places.

Reply:
> the main on is a bit down the line. Distinguishing what to display as the top asset in the hierachy widget appearing under the tabs. If you have old style silos, the top asset in the tree view is the one with visualParent:None, if you're however using that as tab , it should be the next one in the hierarchy. So if you allow mixed type in the project, it goes a bit down the rabit hole. Isn't it more trouble than it's worth?

I see what you mean - but in essence only the data differs but with the code it should still behave exactly the same to the end user. It shouldn't break, even if only later you create the actual parent asset for it. :) The only danger might be that you have a `film` asset that is not supposed to be a silo. So say you have asset "film" with its silo set to "film". ;)

Basically, in backwards compatibility all it means is that:
- When something has a silo set and *no* visual parent. It's parented under a silo tab.
- When something has *no* silo set and *no* visual parent. It is the silo asset itself - it is the tab.
- Any asset with a visualParent is never a silo and always only visible under the parent.

---

- Odd case: When something has a silo set and *no* visual parent but there is an asset for that silo then it is actually "parented" to the tab - it just doesn't have the `visualParent` data set to it. In the data it might look funny, but functionally it's the same as the above cases.